src/hotspot/share/code/compiledMethod.cpp

```diff
@@ -466,70 +428,33 @@ class CheckClass : public MetadataClosure {
-
-    if (cichk_metdata->is_loader_alive()) {
-      return true;
-    }
-  } else {
-    Metadata* ic_metdata = ic->cached_metadata();
-    if (ic_metdata != nullptr) {
-      if (ic_metdata->is_klass()) {
-        if (((Klass*)ic_metdata)->is_loader_alive()) {
-          return true;
-        }
-      } else if (ic_metdata->is_method()) {
-        Method* method = (Method*)ic_metdata;
-        assert(!method->is_old(), "old method should have been cleaned");
-        if (method->method_holder()->is_loader_alive()) {
-          return true;
-        }
-      } else {
-        ShouldNotReachHere();
-      }
-    } else {
-      // This inline cache is a megamorphic vtable call. Those ICs never hold
-      // any Metadata and should therefore never be cleaned by this function.
-      return true;
-    }
-  }
-
-  return ic->set_to_clean();
+static void clean_ic_if_metadata_is_dead(CompiledIC *ic) {
+  ic->clean_metadata();
 }
 
 // Clean references to unloaded nmethods at addr from this one, which is not unloaded.
-template <class CompiledICorStaticCall>
-static bool clean_if_nmethod_is_unloaded(CompiledICorStaticCall *ic, address addr, CompiledMethod* from,
+static void clean_if_nmethod_is_unloaded(CompiledIC *ic, CompiledMethod* from,
                                          bool clean_all) {
-  CodeBlob *cb = CodeCache::find_blob(addr);
-  CompiledMethod* nm = (cb != nullptr) ? cb->as_compiled_method_or_null() : nullptr;
-  if (nm != nullptr) {
-    // Clean inline caches pointing to bad nmethods
-    if (clean_all || !nm->is_in_use() || nm->is_unloading() || (nm->method()->code() != nm)) {
-      if (!ic->set_to_clean(!from->is_unloading())) {
-        return false;
-      }
-      assert(ic->is_clean(), "nmethod " PTR_FORMAT "not clean %s", p2i(from), from->method()->name_and_sig_as_C_string());
-    }
+  CodeBlob* cb = CodeCache::find_blob(ic->destination());
+  if (!cb->is_compiled()) {
+    return;
+  }
+  CompiledMethod* cm = cb->as_compiled_method();
+  if (clean_all || !cm->is_in_use() || cm->is_unloading() || cm->method()->code() != cm) {
+    ic->set_to_clean();
   }
-  return true;
-}
-
-static bool clean_if_nmethod_is_unloaded(CompiledIC *ic, CompiledMethod* from,
-                                         bool clean_all) {
-  return clean_if_nmethod_is_unloaded(ic, ic->ic_destination(), from, clean_all);
 }
 
-static bool clean_if_nmethod_is_unloaded(CompiledStaticCall *csc, CompiledMethod* from,
+static void clean_if_nmethod_is_unloaded(CompiledDirectCall *cdc, CompiledMethod* from,
                                          bool clean_all) {
-  return clean_if_nmethod_is_unloaded(csc, csc->destination(), from, clean_all);
+  CodeBlob* cb = CodeCache::find_blob(cdc->destination());
+  if (!cb->is_compiled()) {
+    return;
+  }
+  CompiledMethod* cm = cb->as_compiled_method();
+  if (clean_all || !cm->is_in_use() || cm->is_unloading() || cm->method()->code() != cm) {
+    cdc->set_to_clean();
+  }
```

Comment:
Maybe a single `clean_if_nmethod_is_unloaded()` with the destination address as a parameter would avoid the code duplication with the other variant; otherwise another static helper function would be great.

Reply:
Good point. I'll try on a template thing.